pkcs11/pkcs11/src/session.rs

```diff
@@ -266,6 +279,356 @@ impl std::fmt::Display for FindObjectsError {
+                    ulValueLen: true_size,
+                },
+            ];
+
+            if let Some(label) = label {
+                key_template.push(pkcs11_sys::CK_ATTRIBUTE_IN {
+                    r#type: pkcs11_sys::CKA_LABEL,
+                    pValue: label.as_ptr() as _,
+                    ulValueLen: std::convert::TryInto::try_into(label.len())
+                        .expect("usize -> CK_ULONG"),
+                });
+            }
+
+            let (mechanism, key_type) = match usage {
+                KeyUsage::Aes => {
+                    key_template.push(pkcs11_sys::CK_ATTRIBUTE_IN {
+                        r#type: pkcs11_sys::CKA_DECRYPT,
+                        pValue: r#true,
+                        ulValueLen: true_size,
+                    });
+                    key_template.push(pkcs11_sys::CK_ATTRIBUTE_IN {
+                        r#type: pkcs11_sys::CKA_ENCRYPT,
+                        pValue: r#true,
+                        ulValueLen: true_size,
+                    });
+
+                    (pkcs11_sys::CKM_AES_KEY_GEN, pkcs11_sys::CKK_AES)
+                }
+
+                KeyUsage::Hmac => {
+                    key_template.push(pkcs11_sys::CK_ATTRIBUTE_IN {
+                        r#type: pkcs11_sys::CKA_SIGN,
+                        pValue: r#true,
+                        ulValueLen: true_size,
+                    });
+
+                    (
+                        pkcs11_sys::CKM_GENERIC_SECRET_KEY_GEN,
+                        pkcs11_sys::CKK_GENERIC_SECRET,
+                    )
+                }
+            };
+
+            key_template.push(pkcs11_sys::CK_ATTRIBUTE_IN {
+                r#type: pkcs11_sys::CKA_KEY_TYPE,
+                pValue: &key_type as *const _ as _,
+                ulValueLen: std::convert::TryInto::try_into(std::mem::size_of_val(&key_type))
+                    .expect("usize -> CK_ULONG"),
+            });
+
+            let mechanism = pkcs11_sys::CK_MECHANISM_IN {
+                mechanism,
+                pParameter: std::ptr::null(),
+                ulParameterLen: 0,
+            };
+
+            let mut key_handle = pkcs11_sys::CK_INVALID_OBJECT_HANDLE;
+            let result = (self.context.C_GenerateKey)(
+                self.handle,
+                &mechanism,
+                key_template.as_ptr() as _,
+                std::convert::TryInto::try_into(key_template.len()).expect("usize -> CK_ULONG"),
+                &mut key_handle,
+            );
+            if result != pkcs11_sys::CKR_OK {
+                return Err(GenerateKeyError::GenerateKeyFailed(result));
+            }
+            if key_handle == pkcs11_sys::CK_INVALID_OBJECT_HANDLE {
+                return Err(GenerateKeyError::GenerateKeyDidNotReturnHandle);
+            }
+
+            Ok(crate::Object::new(self, key_handle))
+        }
+    }
+}
+
+/// An error from generating a key pair.
+#[derive(Debug)]
+#[allow(clippy::pub_enum_variant_names)]
```

Comment:
Not sure why this allow is needed?
https://rust-lang.github.io/rust-clippy/master/#pub_enum_variant_names

Reply:
It isn't. It came along for the right with the existing enum I copied it from, which also didn't need it. (It in turn probably copied it from the `object.rs` enums, which do need it.)

Fixed.